pkg/service/backup/restore_worker_data.go

```diff
@@ -0,0 +1,667 @@
+		return err
+	}
+
+	pr.AgentJobID = jobID
+	w.insertRunProgress(ctx, pr)
+
+	return nil
+}
+
+// newRunProgress creates RestoreRunProgress by assembling batch and starting download to host's upload dir.
+func (w *restoreWorker) newRunProgress(ctx context.Context, run *RestoreRun, target RestoreTarget, h *restoreHost, dstDir, srcDir string,
+) (*RestoreRunProgress, error) {
+	if err := w.checkAvailableDiskSpace(ctx, hostInfo{IP: h.Host}); err != nil {
+		return nil, errors.Wrap(err, "validate free disk space")
+	}
+
+	takenIDs := w.chooseIDsForBatch(ctx, h.Shards, target.BatchSize)
+	if ctx.Err() != nil {
+		w.returnBatchToPool(takenIDs)
+		return nil, ctx.Err()
+	}
+	if takenIDs == nil {
+		return nil, nil //nolint: nilnil
+	}
+
+	w.Logger.Info(ctx, "Created new batch",
+		"host", h.Host,
+		"sstable_id", takenIDs,
+	)
+
+	jobID, err := w.startDownload(ctx, h.Host, dstDir, srcDir, takenIDs)
+	if err != nil {
+		w.returnBatchToPool(takenIDs)
+		return nil, err
+	}
+
+	pr := &RestoreRunProgress{
+		ClusterID:    run.ClusterID,
+		TaskID:       run.TaskID,
+		RunID:        run.ID,
+		ManifestPath: run.ManifestPath,
+		Keyspace:     run.Keyspace,
+		Table:        run.Table,
+		Host:         h.Host,
+		AgentJobID:   jobID,
+		SSTableID:    takenIDs,
+	}
+
+	w.insertRunProgress(ctx, pr)
+
+	return pr, nil
+}
+
+// startDownload creates rclone job responsible for downloading SSTables with given IDs.
+func (w *restoreWorker) startDownload(ctx context.Context, host, dstDir, srcDir string, sstableIDs []string) (int64, error) {
+	batch := w.batchFromIDs(sstableIDs)
+
+	jobID, err := w.Client.RcloneCopyPaths(ctx, host, dstDir, srcDir, batch)
+	if err != nil {
+		return 0, errors.Wrap(err, "download batch to upload dir")
+	}
+
+	w.Logger.Info(ctx, "Started downloading files",
+		"host", host,
+		"job_id", jobID,
+		"batch", batch,
+	)
+
+	return jobID, nil
+}
+
+// validateRunningJob checks if rclone job described in RestoreRunProgress is in running or success state.
+func (w *restoreWorker) validateRunningJob(ctx context.Context, pr *RestoreRunProgress) (bool, error) {
+	job, err := w.Client.RcloneJobProgress(ctx, pr.Host, pr.AgentJobID, w.Config.LongPollingTimeoutSeconds)
+	if err != nil {
+		return false, err
+	}
+
+	if job.Status == string(scyllaclient.JobSuccess) || job.Status == string(scyllaclient.JobRunning) {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// chooseIDsForBatch returns slice of IDs of SSTables that the batch consists of.
+func (w *restoreWorker) chooseIDsForBatch(ctx context.Context, shards uint, size int) []string {
+	var (
+		batchSize = size * int(shards)
+		takenIDs  []string
+	)
+
+	// All restore hosts are trying to get IDs for batch from the pool.
+	// Pool is closed after the whole table has been restored.
+
+	// Take at most batchSize IDs
+	for i := 0; i < batchSize; i++ {
+		select {
+		case id, ok := <-w.bundleIDPool:
+			if !ok {
+				return takenIDs
+			}
+			takenIDs = append(takenIDs, id)
+			if ctx.Err() != nil {
```

Comment:
Why ?

Do you want to handle the situation when the context is cancelled ?
If so, then take a look on this snippet https://go.dev/play/p/rGEzKDEkaXr

It is enough to add `case <- ctx.Done()` as the first check of select.
When the context is cancelled, then the ctx.Done() channel is closed what directly means that `case <-ctx.Done()` is always chosen from select if it's at the first place.

Reply:
> When the context is cancelled, then the ctx.Done() channel is closed what directly means that case <-ctx.Done() is always chosen from select if it's at the first place.

Are you sure that selects work like that in go? Take a look on this snippet https://go.dev/play/p/FtFclYmptAx
Here all channels have values inside of them, but select isn't always choosing the first case - it does that randomly.

Reply:
In my code I wanted to make sure that returned takenIDs are always accurate - even in case of context cancellation. That's why even if I take sth out of the channel when context is cancelled, I still append it to the takenIDs so that it does not get lost in the process.